toc/rfc/rfc-draft-generic-per-route-features.md

```diff
@@ -0,0 +1,148 @@
+# Meta
+[meta]: #meta
+- Name: Generic Per-Route Features
+- Start Date: 2024-01-08
+- Author(s): @maxmoehl, @domdom82
+- Status: Draft <!-- Acceptable values: Draft, Approved, On Hold, Superseded -->
+- RFC Pull Request: (fill in with PR link after you submit it)
+
+## Summary
+
+## Problem
+
+Implementing per-route features in Cloud Foundry is a tedious process because it requires changes to
+multiple components in multiple working groups. Most components will only need to be made aware of
+new fields without adding any significant logic but the overhead of making the change still exists.
+
+On the other side are the components that actually need to be aware of the change. Mostly the
+routing components (gorouter, maybe route-emitter / -registrar), the API (cloud controller) and
+UIs (CLI).
+
+## Proposal
+
+To reduce the overall overhead of implementing new per-route features we propose to add a generic
+key-value map to the definition of a route across all components. All components handling the map
+SHOULD ignore the contents of the map except gorouter, cloud controller and cf CLI which MUST handle
+the map and its values accordingly. Other components MAY limit the size of the map or size of keys /
+values for technical reasons.
+
+The following constraints apply (types are as specified in [RFC 8259](https://rfc-editor.org/rfc/rfc8259)):
+* The map MUST be representable as an object.
+* The map MUST only use strings as keys.
+* The map MUST only use numbers, strings and the literal values `true` and `false` as values.
+
+### Required Changes
+
+#### CLI
+
+There MAY be either individual command line flags or a single flag which supports generic key-value
+pairs for the `map-route` sub-command.
+
+#### Cloud Controller
+
+The API MUST specify and implement a new field to carry the map inside the route object. The
+API MUST implement per-field validations as features of the map are specified. The route MUST be
+rejected with an appropriate error message if the provided map is invalid and MUST be passed on
+as-is otherwise. The API also MUST store the map as a generic key-value map to ensure changes to the
+map do not require a change to the database schema.
+
+Support for a generic map MUST be added to the manifest. The updated routes section in a manifest
+MAY look like this:
+
+```yml
+version: 1
+applications:
+- name: test
+  routes:
+  - route: example.com
+    options:
```

Comment:
If the route already exists without these options would you expect it to be updated?

There hasn't really been much updating of routes before so you might want to mention that this endpoint would be enhanced:
https://v3-apidocs.cloudfoundry.org/version/3.159.0/index.html#update-a-route

And probably worth mentioning https://v3-apidocs.cloudfoundry.org/version/3.159.0/index.html#create-a-route as well. `cf create-route` and `cf map-route` (if the route doesn't already exist) both call that one.

Reply:
Good point! Yes, we expect those options to be modifiable via updates.

Reply:
I'm not that familiar with the CC API, are such `PATCH` calls usually additive or do they replace the object?

Reply:
Do you think this should be included in the RFC? I'd probably leave such details up to the implementers.

I've also added the relevant API endpoints to the RFC.

Reply:
I don't think it needs to get into a great amount of detail, but maybe a line saying that this config in the manifest will either create a new route with these options or update an existing route to have these options. Main reason I think that's important is because the current manifest logic is somewhat inconsistent (e.g. removing env vars from a manifest wont remove them from an app).

Speaking of that. If person A created a route ahead of time (`cf create-route`) with `options`, but someone pushed an app with a manifest that just mentioned the route, but without options, would you expect the options to be removed from the route? Or should that require a more intentional action?

Reply:
I've been thinking about this for the past week and can't really decide what feels "right". But I'd be in favour of having to explicitly un-set the options. E.g. if you don't provide options they remain untouched and if you submit an empty map `{}` they will be removed.

Given:

```yaml
version: 1
applications:
- name: test
  routes:
  - route: a.example.com
    options:
      loadbalancing-algorithm: round-robin
      trim-path: true
  - route: b.example.com
    protocol: http2
  - route: b.example.com
    protocol: http2
    options: {}
```

Assuming all of the routes exist with some options already set:

* `a` would replace any existing options map with the one provided (potentially dropping what was set for, e.g. `session-cookie`).
* `b` would not change the options, if they exist, they stay.
* `c` would remove any previously attached options.

Reply:
I would be in favour of a "more" declarative approach here. Talking to users they see it as an CF issue that env vars aren't removed when they are not present in the manifest any more or service bindings are not touched, when they are removed from the manifest. In general we should strive to have the "desired state truth" defined in the app manifest.

Reply:
Agreed, but I wouldn't want to introduce a new feature that behaves differently to all the existing ones, that would probably cause even more confusion.

Reply:
I'd be interested to know how it is implemented today?
If I do first
```
cf map-route my-app example.com --hostname myhost --app-protocol http2
```
then do
```
cf push
```
with a manifest saying
```
applications:
- name: my-app
  routes:
  - route: myhost.example.com
```
so no `protocol` key was defined in the manifest.
What does it do? Does it keep `http2` or overwrite with the default of `http1`?

Whatever we do today, it should be the same with the `options` field IMO.

Reply:
I agree with the suggestion of @domdom82 to have this consistent with the current implementation of the `http2` option. My intention was to provide feedback from CF users and motivate for a more declarative approach if it is possible and still keep things consistent.

Reply:
I tried this out. First I defined the route for my app with:
```
cf map-route cf-nodejs-bv-test <domain> --hostname cf-nodejs-bv-test-zany-chimpanzee-ra --app-protocol http2
```
Afterwards, I added that rout to my manifest with:
```
---
applications:
- name: cf-nodejs-bv-test
  memory: 1GB
  instances: 1
  random-route: true
  routes:
  - route: cf-nodejs-bv-test-zany-chimpanzee-ra.<domain>
```

Then I pushed the app:

```
...
Updating with these attributes...
  ---
  applications:
  - name: cf-nodejs-bv-test
    instances: 1
    memory: 1GB
+   random-route: true
    routes:
+   - route: cf-nodejs-bv-test-zany-chimpanzee-ra.<domain>
-   - protocol: http2
-     route: cf-nodejs-bv-test-zany-chimpanzee-ra.<domain>
...
```

The push diff states that the `http2` protocal will be removed but checking with the CF APIs the protocol is still `http2`
```
cf curl "/v3/routes?app_guids=3f3bbd62-d46a-4e55-871e-f17a551a6aac" | jq .
...
{
      "guid": "322a5fe3-3f93-4ecb-9e9c-b1103c327643",
      "created_at": "2024-04-08T08:40:22Z",
      "updated_at": "2024-04-08T08:40:22Z",
      "protocol": "http",
      "host": "cf-nodejs-bv-test-zany-chimpanzee-ra",
      "path": "",
      "port": null,
      "url": "cf-nodejs-bv-test-zany-chimpanzee-ra.<domain>",
      "destinations": [
        {
          "guid": "f6482e06-736b-4555-87a5-22cfa68b7848",
          "app": {
            "guid": "3f3bbd62-d46a-4e55-871e-f17a551a6aac",
            "process": {
              "type": "web"
            }
          },
          "weight": null,
          "port": 8080,
          "protocol": "http2"
        }
      ],
```
I'm not sure whether this is a bug in CAPI or the expected behaviour. @tcdowney or @stephanme could you please comment on that?

Reply:
I think this (indeed surprising) behaviour is correct according to the API spec: https://v3-apidocs.cloudfoundry.org/#apply-a-manifest-to-a-space

> Apply changes specified in a manifest to the named apps and their underlying processes. The apps must reside in the space. These changes are additive and will not modify any unspecified properties or remove any existing environment variables, routes, or services.

Since the protocol property was not specified, it remains unchanged which results in http2. The diff output could be considered as buggy.

Disclaimer: I don't understand the coding in every detail and I could not find a matching test case that deals with this protocol case. I will raise this in the CAPI Open Office Hour.

Reply:
We discussed this during the TOC meeting today and we should go for consistent behaviour as documented in the APIs.

Reply:
I've added a paragraph to capture that:

> The overall goal MUST be to implement the new fields in a way that fits in with the current behavior of the cloud controller API. It is up to the maintainers / implementers to decide on the exact behavior.

Reply:
I don't have a terribly strong opinion here, but I think we should go with what's consistent for now and provide a way for users to explicitly unset the properties in the manifest.

I know that @Gerg has been thinking about what a truly declarative "version 2" of the CF manifest my look/behave like and I think that would be an opportunity to incorporate some of that feedback. He has a doc somewhere, but I think it was made private due to some Google doc migrations. I'll let him chime in when he's available / let him know about this.

Reply:
> we should go for consistent behaviour as documented in the APIs.

I agree. I also remember being confused about this behavior regarding readiness health check manifest properties. It's odd, but that is how the app manifest currently works. I think changing it is out of scope for this RFC.